### PR TITLE
change setpgid to setsid

### PR DESCRIPTION
**Pull request by kentwills**
Test failure without the patch:

```
$ ./tests/test-tty /nail/home/rkwills/trees/yelp/dumb-init/dumb-init
Traceback (most recent call last):
  File "./tests/test-tty", line 42, in <module>
    main()
  File "./tests/test-tty", line 40, in main
    test_tty(argv[1])
  File "./tests/test-tty", line 35, in test_tty
    parent(fd)
  File "./tests/test-tty", line 25, in parent
    assert output == '3\n2\n1\n', repr(output)
AssertionError: "Timeout: aborting command ``/nail/home/rkwills/trees/yelp/dumb-init/dumb-init'' with signal 9\n"
```

In short, using setpgid makes the posix tty system think our subprocess is a backgrounded shell job, which it isn't.  A backgrounded shell job gets SIGTTOU whenever it prints, which puts it into a paused state, which is killable only with SIGKILL.

Putting it into another session entirely represents it more accurately; it doesn't compete with any shell jobs for tty output.

### CODE REVIEW

**Comment by kentwills** [comment]
@chriskuehl @bukzor

File: dumb-init.c [hunk 1]
```diff
@@ -110,8 +110,8 @@ int main(int argc, char *argv[]) {
 
     if (child == 0) {
         if (use_process_group) {
-            pid_t result = setpgid(0, 0);
-            if (result != 0) {
+            pid_t result = setsid();
+            if (result == -1) {
                 fprintf(
                     stderr,
                     "Unable to create process group (errno=%d %s). Exiting.\n",
```
Comment (chriskuehl) [inline dumb-init.c]: update this? (no longer does anything involving process groups)

Comment (chriskuehl) [inline dumb-init.c]: Also update line 123? (GitHub won't let me put a note on that line for some reason...)

`DEBUG("Set process group ID of child to its own PID.\n");`

Comment (chriskuehl) [inline dumb-init.c]: Actually, would it also make sense to just rename the environment variable to `DUMB_INIT_SESSION` or something?

File: tests/test-tty [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python
```
Comment (chriskuehl) [inline tests/test-tty]: Can you add some kind of docs to this test? It's not clear to me at all what is being tested here.

File: tests/test-tty [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+def child(dumb_init):
+    from os import execvp
+    execvp('timeout', ('timeout', '1', dumb_init, 'tac'))
+
+
+def parent(fd):
+    ttyflags(fd)
+    from os import read, write
+    assert write(fd, b'1\n2\n3\n') == 6
```
Comment (chriskuehl) [inline tests/test-tty]: I have no idea why this assertion should be true. A few comments would be really helpful.

Comment (chriskuehl) [inline tests/test-tty]: oh, never mind, it's clear now.

I think docstrings on each of these methods would be really helpful though, I don't really know what `parent(fd)` means or is testing.

**Comment by chriskuehl** [comment]
This looks good. A couple things:

* Can you confirm if `make itest` is still passing? It requires Docker so is not run by Travis (I'm working on moving to CircleCI instead)
* The README, help output, environment variable, and possibly Debian control file (`debian/control`) should probably be updated.

**Comment by chriskuehl** [comment]
`make itest` failed for me (it didn't get past `make builddeb-docker`):
https://gist.github.com/anonymous/cf7a9fa93f9ae62e2f9e

**Comment by kentwills** [comment]
@chriskuehl Added your change, should be ready to merge.

File: test [hunk 4]
```diff
@@ -11,15 +11,18 @@ fi
 echo "Running with dumb-init at '$dumb_init_bin'"
 
 run_tests() {
-    export DUMB_INIT_PROCESS_GROUP
-    for DUMB_INIT_PROCESS_GROUP in 0 1; do
+    ./test-tty "$dumb_init_bin"
+    return 0
```
Comment (chriskuehl) [inline test]: derp

File: tests/test-pgroup [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash -eux
-# dumb-init should proxy signals to a process group rooted at its child when
+# dumb-init should proxy signals to a session rooted at its child when
```
Comment (chriskuehl) [inline tests/test-pgroup]: should this test be renamed?

File: tests/test-tty (hunk 2 above)
Comment (chriskuehl) [inline tests/test-tty]: thanks!

File: tests/test-tty [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    T.tcsetattr(fd, T.TCSANOW, attrs)
+
+
+def tac(dumb_init):
+    """
+    run tac. if it fails to complete in 1 second send SIGKILL and exit with an
+    error.
+    """
+    from os import execvp
+    execvp('timeout', ('timeout', '1', dumb_init, 'tac'))
```
Comment (chriskuehl) [inline tests/test-tty]: hackathon idea: write `taclol` to complement `lolcat`

**Comment by chriskuehl** [comment]
LGTM besides a couple minor nits above

File: .dockerignore [hunk 7]
```diff
@@ -0,0 +1,2 @@
+virtualenv_run
+.git
```
Comment (chriskuehl) [inline .dockerignore]: don't think we need dockerignore, we only mount into the image, not ADD/COPY

Comment (chriskuehl) [inline .dockerignore]: per IRC, we should probably keep this.

I think `virtualenv_run` should probably me changed to `.tox` though?

File: tests/test-pgroup (hunk 5 above)
Comment (bukzor) [inline tests/test-pgroup]: test-setsid i suppose

File: tests/test-tty (hunk 6 above)
Comment (bukzor) [inline tests/test-tty]: fun fact: lol is lol, backwards

**Comment by kentwills** [comment]
@chriskuehl updated

**Comment by chriskuehl** [comment]
shipit



assuming good tests (and they look fine so far, minus issue #15 flakiness)

**Pull request merged**